### PR TITLE
Do not collect a stack sample/callstack on uretprobes

### PR DESCRIPTION
**Pull request by dpallotti**

### CODE REVIEW

File: OrbitLinuxTracing/PerfEvent.cpp [hunk 1]
```diff
@@ -26,19 +26,19 @@ void StackSamplePerfEvent::accept(PerfEventVisitor* visitor) {
...
   visitor->visit(this);
 }
 
-void UretprobePerfEvent::accept(PerfEventVisitor* visitor) {
+void UretprobesPerfEvent::accept(PerfEventVisitor* visitor) {
   visitor->visit(this);
 }
 
-void UretprobePerfEventWithStack::accept(PerfEventVisitor* visitor) {
+void UretprobesWithStackPerfEvent::accept(PerfEventVisitor* visitor) {
```
Comment (ronaldfw) [inline OrbitLinuxTracing/PerfEvent.cpp]: Why is this still needed?

File: OrbitLinuxTracing/PerfEvent.h [hunk 2]
```diff
@@ -256,13 +256,13 @@ class UprobePerfEventWithStack
...
 
-class UretprobePerfEvent : public AbstractUprobePerfEvent<perf_empty_record> {
+class UretprobesPerfEvent : public AbstractUprobesPerfEvent<perf_empty_record> {
  public:
   void accept(PerfEventVisitor* visitor) override;
 };
 
-class UretprobePerfEventWithStack
-    : public AbstractUprobePerfEvent<perf_record_with_stack> {
+class UretprobesWithStackPerfEvent
```
Comment (ronaldfw) [inline OrbitLinuxTracing/PerfEvent.h]: Why is this still needed?

File: OrbitLinuxTracing/UprobesUnwindingVisitor.cpp [hunk 3]
```diff
@@ -100,6 +134,17 @@ UprobesCallstackManager::ProcessUretprobesCallstack(
   return full_callstack;
 }
 
+void UprobesCallstackManager::ProcessUretprobesWithoutCallstack(pid_t tid) {
```
Comment (ronaldfw) [inline OrbitLinuxTracing/UprobesUnwindingVisitor.cpp]: Can the function above (ProcessUretprobesCallstack) be deleted?

File: OrbitLinuxTracing/UprobesUnwindingVisitor.cpp [hunk 4]
```diff
@@ -2,6 +2,40 @@
...
+void UprobesFunctionCallManager::ProcessUprobes(pid_t tid,
+                                                uint64_t function_address,
+                                                uint64_t begin_timestamp) {
+  auto& tid_timer_stack = tid_timer_stacks_[tid];
+  tid_timer_stack.emplace(function_address, begin_timestamp);
+}
+
+std::optional<FunctionCall> UprobesFunctionCallManager::ProcessUretprobes(
+    pid_t tid, uint64_t end_timestamp) {
+  if (tid_timer_stacks_.count(tid)) {
```
Comment (ronaldfw) [inline OrbitLinuxTracing/UprobesUnwindingVisitor.cpp]: Would be easier to read if we "early out" here. Then you don't have to indent most of the function body. 

if (tid_timer_stacks.count(tid) == 0) {
  return std::optional<FunctionCall>();
}

// Now we know what tid_timer_stacks.count(tid) > 0

auto& tid_timer_stack = ...

File: OrbitLinuxTracing/UprobesUnwindingVisitor.cpp [hunk 5]
```diff
@@ -2,6 +2,40 @@
...
+    auto& tid_timer_stack = tid_timer_stacks_.at(tid);
+
+    // This is because we erase the stack for this thread as soon as it becomes
+    // empty.
+    assert(!tid_timer_stack.empty());
+
+    if (!tid_timer_stack.empty()) {
```
Comment (ronaldfw) [inline OrbitLinuxTracing/UprobesUnwindingVisitor.cpp]: Why the "if" and the "assert" with the same condition?

Comment (ronaldfw) [inline OrbitLinuxTracing/UprobesUnwindingVisitor.cpp]: If you keep the "if" despite the "assert", also early out here:

if (tid_timer_stack.empty()) {
  return std::optional<FunctionCall>{};
}

// Now comes the real body of the function.

File: OrbitLinuxTracing/UprobesUnwindingVisitor.cpp [hunk 6]
```diff
@@ -140,19 +182,35 @@ void UprobesUnwindingVisitor::visit(UprobePerfEventWithStack* event) {
...
-void UprobesUnwindingVisitor::visit(UretprobePerfEventWithStack* event) {
-  FunctionEnd function_end{event->TID(), event->GetFunction()->VirtualAddress(),
-                           event->Timestamp()};
-  if (listener_ != nullptr) {
-    listener_->OnFunctionEnd(function_end);
+void UprobesUnwindingVisitor::visit(UretprobesPerfEvent* event) {
+  std::optional<FunctionCall> function_call =
+      function_call_manager_.ProcessUretprobes(event->TID(),
+                                               event->Timestamp());
+  if (function_call.has_value()) {
```
Comment (ronaldfw) [inline OrbitLinuxTracing/UprobesUnwindingVisitor.cpp]: Do both conditions in the same if: 

if (function_call.has_value() && listener_ != nullptr) {
 ...
}

File: OrbitLinuxTracing/UprobesUnwindingVisitor.cpp [hunk 7]
```diff
@@ -140,19 +182,35 @@ void UprobesUnwindingVisitor::visit(UprobePerfEventWithStack* event) {
...
+  }
+
+  callstack_manager_.ProcessUretprobesWithoutCallstack(event->TID());
+}
+
+void UprobesUnwindingVisitor::visit(UretprobesWithStackPerfEvent* event) {
+  std::optional<FunctionCall> function_call =
+      function_call_manager_.ProcessUretprobes(event->TID(),
+                                               event->Timestamp());
+  if (function_call.has_value()) {
```
Comment (ronaldfw) [inline OrbitLinuxTracing/UprobesUnwindingVisitor.cpp]: Same here.

File: OrbitLinuxTracing/UprobesUnwindingVisitor.h [hunk 8]
```diff
@@ -23,6 +26,34 @@ namespace LinuxTracing {
 // of instrumented functions. When we have a callstack broken because of
 // uretprobes we can then rebuild the missing part by joining together the parts
 // on the stack of callstacks associated with that thread.
+// TODO: Make this more robust to losing uprobes or uretprobes events (loss of
+//  uretprobes events should be rare if they don't come with a stack sample).
+//  Start by passing the function_address to ProcessUretprobes as well and for a
```
Comment (ronaldfw) [inline OrbitLinuxTracing/UprobesUnwindingVisitor.h]:
```suggestion
//  Start by passing the function_address to ProcessUretprobes as well for a
```

File: OrbitLinuxTracing/UprobesUnwindingVisitor.h [hunk 9]
```diff
@@ -23,6 +26,34 @@ namespace LinuxTracing {
...
+  UprobesFunctionCallManager(UprobesFunctionCallManager&&) = default;
+  UprobesFunctionCallManager& operator=(UprobesFunctionCallManager&&) = default;
+
+  void ProcessUprobes(pid_t tid, uint64_t function_address,
+                      uint64_t begin_timestamp);
+  std::optional<FunctionCall> ProcessUretprobes(pid_t tid,
+                                                uint64_t end_timestamp);
+
+ private:
+  // This map keeps the stack of the dynamically-instrumented functions entered.
```
Comment (ronaldfw) [inline OrbitLinuxTracing/UprobesUnwindingVisitor.h]: This type is pretty unreadable. What do the uint64_t types represent? What to the pairs represent? These should be properly named types (structs).

File: OrbitLinuxTracing/UprobesUnwindingVisitorTest.cpp [hunk 10]
```diff
@@ -7,6 +7,94 @@
...
+  EXPECT_EQ(processed_function_call.value().GetDepth(), 0);
+}
+
+TEST(UprobesFunctionCallManager, TwoNestedUprobesAndAnotherUprobe) {
+  constexpr pid_t tid = 42;
+  std::optional<FunctionCall> processed_function_call;
+  UprobesFunctionCallManager function_call_manager;
+
+  function_call_manager.ProcessUprobes(tid, 100, 1);
+
```
Comment (ronaldfw) [inline OrbitLinuxTracing/UprobesUnwindingVisitorTest.cpp]: Remove this line?

File: OrbitLinuxTracing/UprobesUnwindingVisitorTest.cpp [hunk 11]
```diff
@@ -7,6 +7,94 @@
...
+}
+
+TEST(UprobesFunctionCallManager, TwoUprobesDifferentThreads) {
+  constexpr pid_t tid = 42;
+  constexpr pid_t tid2 = 111;
+  std::optional<FunctionCall> processed_function_call;
+  UprobesFunctionCallManager function_call_manager;
+
+  function_call_manager.ProcessUprobes(tid, 100, 1);
+
```
Comment (ronaldfw) [inline OrbitLinuxTracing/UprobesUnwindingVisitorTest.cpp]: Remove this line?

File: OrbitLinuxTracing/UprobesUnwindingVisitorTest.cpp [hunk 12]
```diff
@@ -112,6 +200,34 @@ TEST(UprobesCallstackManager, OneUprobe) {
                   TestCallstackToStringPairVector(expected_cs)));
 }
 
+TEST(UprobesCallstackManager, OneUprobeNoCallstackOnUretprobe) {
+  constexpr pid_t tid = 42;
+  std::vector<unwindstack::FrameData> unwound_cs, expected_cs, processed_cs;
+  UprobesCallstackManager callstack_manager{};
+  // Begin FUNCTION().
```
Comment (ronaldfw) [inline OrbitLinuxTracing/UprobesUnwindingVisitorTest.cpp]: I don't understand what this comment here means. Can you please add more explanation here?

File: OrbitLinuxTracing/PerfEvent.cpp (hunk 1 above)
Comment (dpallotti) [reply OrbitLinuxTracing/PerfEvent.cpp]: UprobesPerfEvent is also not used. It was for consistency. I'll remove both.

File: OrbitLinuxTracing/UprobesUnwindingVisitor.cpp (hunk 7 above)
Comment (dpallotti) [reply OrbitLinuxTracing/UprobesUnwindingVisitor.cpp]: Done.

File: OrbitLinuxTracing/UprobesUnwindingVisitor.cpp (hunk 3 above)
Comment (dpallotti) [reply OrbitLinuxTracing/UprobesUnwindingVisitor.cpp]: Done.

File: OrbitLinuxTracing/UprobesUnwindingVisitor.cpp (hunk 6 above)
Comment (dpallotti) [reply OrbitLinuxTracing/UprobesUnwindingVisitor.cpp]: Done.

File: OrbitLinuxTracing/UprobesUnwindingVisitorTest.cpp (hunk 12 above)
Comment (dpallotti) [reply OrbitLinuxTracing/UprobesUnwindingVisitorTest.cpp]: It means that we simulate the call of the function FUNCTION. I'll change the comments.

**Pull request merged**